unicode-org/icu4x
# Add feature slicing to capi

Fixes https://github.com/unicode-org/icu4x/issues/2602


Some things I'm not sure about:

 - Should we have separate datetime/timezone features? The FFI APIs overlap a lot
 - Should we match features to crate names or vague components (I'm currently doing `icu_datetime`, etc, but we could just have `datetime`, especially if it's also enabling stuff like timezones)
 - How should we handle the features test?


For the features test (which will take forever otherwise) I have a couple ideas:

 - Disable it for this crate and write out manual permutations that we care about
 - Add a `always_include` cfg to upstream and always include `default_components`. Run a second test that tests each component individually. This should be sufficient, I think, to ensure that the component CFGs are correct, and we don't need to worry too much about what happens when you include two
 - Add a `linearize` cfg to upstream that lets you specify lists of features that there should be at most one of. I'm not sure if this can be done in a way that also covers `default_components`.

In general I'm fine with limited feature testing for this slicing: if a client complains we can always release patches, and we claim that default_components is the Normal way to interact with icu_capi. We definitely should test all feature combos with `default_components` enabled. I'm leaning towards the second option.
